src/Google.Cloud.Functions.Framework/CloudEventAdapterTData.cs

```diff
@@ -31,22 +33,27 @@ public sealed class CloudEventAdapter<TData> : IHttpFunction where TData : class
         /// <summary>
         /// Constructs a new instance based on the given CloudEvent Function.
         /// </summary>
+        /// <remarks>
+        /// The <paramref name="formatters"/> parameter is a sequence to allow for a diagnostically-helpful exception
+        /// to be thrown if no formatters are available. Only a single formatter is currently used, which will be the last
+        /// one in the sequence to allow for functions to add their own and override the default.
+        /// </remarks>
         /// <param name="function">The CloudEvent Function to invoke.</param>
-        /// <param name="formatter">The CloudEvent formatter to use to deserialize the CloudEvent. This is expected to be a single-type formatter,
-        /// targeting <typeparamref name="TData"/>.</param>
+        /// <param name="formatters">A sequence of CloudEvent formatters to use to deserialize the CloudEvent. See remarks for more details.</param>
         /// <param name="logger">The logger to use to report errors.</param>
-        public CloudEventAdapter(ICloudEventFunction<TData> function, CloudEventFormatter formatter, ILogger<CloudEventAdapter<TData>> logger)
+        public CloudEventAdapter(ICloudEventFunction<TData> function, IEnumerable<CloudEventFormatter> formatters, ILogger<CloudEventAdapter<TData>> logger)
```

Comment:
My initial reaction is that receiving a sequence for null-checking is just a fancy way of implementing a null-forgiving operator. In particular, because we really don't use a sequence at all, just the last element on it, if any. So, I'd probably go with just using the null-forgiving operator.

Still, two other alternatives to consider:

- I think, that if you create a second constructor that receieves no formatter, and only throws he exception, that will be called if there's no formatter registered. I'm pretty sure I've used this before. And then here you continue to receive a single formatter. Granted, having a constructor that only throws is very weird.
- Receive the single non-null formatter, and just let DI throw the exception that it normally throws when it can't resolve a service, and document that very well in a troubleshooting section of the docs.

Reply:
I'd prefer not to go down the second option here - I definitely want us to be able to specify an error message.
My concern over using the null-forgiving operator is that DI might *start* validating that more aggressively later... we're effectively violating the contract and hoping they don't spot it.

I'll investigate the first option though - that sounds promising.

Reply:
I've added a new constructor, and it's working fine. The fourth commit contains that - PTAL.